### ISSUE TITLE
Improve register allocator

### ISSUE THREAD

**Issue by malleusinferni** [rsmw/souvenir#2]
IR-to-opcode translation requires the use of a register allocator. The one currently implemented is probably the worst one ever, because it performs **globally unique allocation**, such that the total number of registers in a VM stack frame is also an upper bound on the total number of scalar variables (including temporaries) that can exist anywhere throughout a compiled program.

This is bad.

The following steps should be taken to make it less bad:

- [ ] Instead of scanning all blocks, restart the scan at each entry point (scene, trap, or prelude) and scan only the blocks accessible from that entry point.
- [ ] Instead of assuming that variables live forever, perform liveness analysis and deallocate registers when the variables they're assigned to are dead.
- [ ] Instead of rejecting programs that have too many live registers, spill excess variables to a structure on the heap, and then retry allocation.

The third step will likely be the most difficult. Choosing which registers to spill is a topic of ongoing research. The first two steps, at least, are critical to writing larger programs with many different scenes.

The situation with flag variables is less dire in the short term, but still poses a problem, because the allocation scheme for flag registers assumes that no flag can live past the end of its own basic block. This prohibits certain optimization passes and unnecessarily complicates others. We ought to be able to use the same allocator for both flags and scalars.

**Comment by malleusinferni** [comment]
Register allocation is no longer as serious a problem as it was with the old interpreter design, both because native compilation has become a lower priority and because it can now be achieved with easier or more reliable techniques. For example, storing the intermediate results of evaluated expressions used to require a temporary variable, but expression evaluation is now guaranteed to be an atomic operation and can be handled by a tree walking evaluator (like we have now) or by "compiling" the AST into nested closures. Any additional efforts will depend on benchmarking in real world use cases and can be delivered by modern JIT backends such as Cranelift or recently developed LLVM APIs.

In short, register allocation efforts are now considered a premature optimization, and this issue can be closed.
